### ISSUE TITLE
HistogramDisplay, plot_stairstep_graph encounters ValueError for range

### ISSUE THREAD

**Issue by gg2** [ARM-DOE/ACT#482]
I'm attempting to generate histograms for some Aerosol Optical Depth data. It is rather sparse, and looks like this:

[nan, nan, 0.04041248, nan, nan, nan, nan, nan, nan, 0.08403611, nan, nan, nan, nan, nan, 0.08443181, 0.05368716, nan, 0.07251927, 0.05726737, nan, 0.04878159, nan, 0.06544831, nan, nan, nan, nan, 0.04621224, nan, nan]

Plotting like so:
```
    var_name = 'aod_be_polyfit_415'
    nbins = 12
    try:
        ts_display = act.plotting.HistogramDisplay(
            nc_data,
            figsize=(19,5.5),
            subplot_shape=(1,),
        )
    except Exception as e:
        print(f"{type(e)} error generating histogram of netCDF data for {var_name}:\n{e}")
    else:
        ts_display.plot_stairstep_graph(var_name, bins=nbins, subplot_index=(0,))
```

It encounters this error:
Traceback (most recent call last):
  File "plot_aodbe_daily_avgs.py", line 335, in <module>
    sys.exit(main())
  File "plot_aodbe_daily_avgs.py", line 320, in main
    ts_display.plot_stairstep_graph(var_name, bins=nbins, subplot_index=(0,))
  File "/home/<username>/.local/lib/python3.9/site-packages/act/plotting/histogramdisplay.py", line 386, in plot_stairstep_graph
    my_hist, bins = np.histogram(
  File "<__array_function__ internals>", line 180, in histogram
  File "/home/<username>/.local/lib/python3.9/site-packages/numpy/lib/histograms.py", line 793, in histogram
    bin_edges, uniform_bins = _get_bin_edges(a, bins, range, weights)
  File "/home/<username>/.local/lib/python3.9/site-packages/numpy/lib/histograms.py", line 426, in _get_bin_edges
    first_edge, last_edge = _get_outer_edges(a, range)
  File "/home/<username>/.local/lib/python3.9/site-packages/numpy/lib/histograms.py", line 323, in _get_outer_edges
    raise ValueError(
ValueError: autodetected range of [nan, nan] is not finite

Since np.histogram accepts a `range` parameter, I assume this error is because it attempts to auto-calculate the range, but the data contains nans and throws that off. I experimented and found if I replace nan with e.g. 1.0, the plotting works. I have to replace all nans.

I imagine there might be various ways to address this.
* Can this be corrected within ACT, so either A) ACT determines the range better than numpy does, ignoring the nans, or B) accepts a range parameter that it will pass to np.histogram?
* Is there a way you might recommend handling this in the data before attempting to plot a histogram? As I noted, I could replace all nans with other values (-9999), but then there'd need to be a way to mask out that value... and might as well just mask the nans.

**Comment by zssherman** [comment]
@gg2 We might need to add a check for it, for now can you np.ma.masked_invalid your range? should just mask the nans then ignore them hopefully

**Comment by gg2** [comment]
Wow! That was fast. I'll try it.

Also, `bins` is not allowed to be `None`, or not provided as a parameter to plot_stairstep_graph() (because it then defaults to `None`.

**Comment by gg2** [comment]
I don't know how to apply the MaskedArray to the underlying Xarray.

```
var_data = nc_data[var_name]
print(type(var_data))
print(type(var_data.values)) # Is the ndarray type that MaskedArray is a variant of.
new_data = np.ma.masked_invalid(var_data.values, copy=True)
print(f"{type(new_data)}:\n{new_data}\n")
# Testing -- creates the same MaskedArray
#new_data = np.ma.masked_invalid(var_data, copy=True)
#print(f"{type(new_data)}:\n{new_data}\n")

data_mask = np.ones(len(var_data.values), dtype=np.int8) # Just all ones, to make an index array, since Xarray is silly about not allowing updates to single values, but also because I wasn't seeing a change with:
#nc_data[var_name].values = new_data # doesn't make a difference
print(f"{data_mask}")
nc_data[var_name].values[data_mask] = new_data

print(f"{var_name} ({type(nc_data[var_name].values)}):\n{nc_data[var_name].values}\n")

# The following successfully updated specific values in previous tests.
#data_nans = np.where(np.isnan(var_data.values))
#print(data_nans)
#var_data[data_nans] = 1.0
```

**Comment by mgrover1** [comment]
@gg2 do you have a sample file we can test with?

**Comment by mgrover1** [comment]
@gg2 one solution would be to drop the missing the values using Xarray's dropna() functionality. It would like something like this:

```python
ds_with_missing_values = act.io.read_netcdf("your_file.nc")

ds_with_missing_values = ds_with_missing_values.dropna("time")

ts_display = act.plotting.HistogramDisplay(
            ds_with_missing_values,
            figsize=(19,5.5),
            subplot_shape=(1,),
        )

ts_display.plot_stairstep_graph("aod_field", bins=12)
```

**Comment by gg2** [comment]
I'll give that a try.
Huh, I thought I searched all past issues for histogram and came up with nothing. Thanks for doing the search for me apparently.

Thank you.

**Comment by mgrover1** [comment]
That was a pull request I worked on this morning! New code :)

**Comment by gg2** [comment]
Just to report back here, I did try dropna. It was simple to apply and worked for the ACT version I'm using (1.1.6).

**Comment by mgrover1** [comment]
Fantastic! Great to hear!

**Comment by zssherman** [comment]
@gg2 We will also be doing a release here shortly, 1.1.9, which should have this built in now, correct @mgrover1 ?

**Comment by mgrover1** [comment]
Yup!

**Comment by gg2** [comment]
I know this one's closed, but I wanted to say "thank you" for making this update.

I ran into another problem with histograms in the same script as originally above. That pointed out I hadn't updated ACT and tried 1.1.9. 1.1.9 seems to generate histograms of data successfully. It looks like dropna() was implemented in ACT better than how I was using it. 

I was simply applying it to the dataset (because I have to hand a dataset to HistogramDisplay). But the problem I encountered today involved histogram data coming in empty, because, it appears, dropna() drops data across the whole dataset for any NaN in a single variable.

And it looks perhaps like that's what you found, too. At the very least, it looks like you only applied dropna() at the level of the DataArray being graphed (in self._get_data() ), not at the DataSet level. (And if the histograms are dropping valid values in a given variable, I'd still have to rig up an auxiliary count to verify. But, I'm guessing, since ACT got past my latest issue, it's applying dropna() appropriately.)

So, thank you, and good job!
